### PR TITLE
Handle multiple aliases in _cat/aliases api

### LINKED ISSUE #23661: cat aliases {name} documented as list instead of string
[cat RestAliasAction](https://github.com/elastic/elasticsearch/blob/6265ef1c1ba1d308bcc28d00dccccac555e33b89/core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java#L43) exposes an endpoint taking `{alias}` which feeds this **single** `alias` into the [GetAliasRequest constructor](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/admin/indices/alias/get/GetAliasesRequest.java#L42) exactly [as documented](https://www.elastic.co/guide/en/elasticsearch/reference/current/cat-alias.html)

However [cat.aliases](https://github.com/elastic/elasticsearch/blob/6265ef1c1ba1d308bcc28d00dccccac555e33b89/rest-api-spec/src/main/resources/rest-api-spec/api/cat.aliases.json#L9-L12) documents it as a `list` instead of a `string`

### PR DESCRIPTION
**Pull request by glefloch**
This pull request aims to handle a list of `alias` in `GetAliasRequest` instead of a single one.

Closes #23661

### CODE REVIEW

**Comment by jasontedor** [comment]
Thanks for the PR, but we'll need tests here; namely, tests that would reveal the broken behavior before your change (which at a superficial glance looks right), and tests that will pass after your change.

**Comment by glefloch** [comment]
Yes for sure, I 'am gonna add that to this pull request

**Comment by glefloch** [comment]
Hi, I looked at the code in order to develop some tests. I'm kind of stuck, I would like to assert the result of the `doCatRequest` wich return a `RestChannelConsumer` object. But I d'ont have access to that class. Are there any available test classes that allow to assert Rest response?

**Comment by jasontedor** [comment]
The simplest option would be to write a REST test; look in rest-api-spec where there tests for similar endpoints that you can base your work from.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java [hunk 1]
```diff
@@ -78,7 +78,8 @@ protected Table getTableWithHeader(RestRequest request) {
         return table;
     }
 
-    private Table buildTable(RestRequest request, GetAliasesResponse response) {
+    // package private for testing
+    Table buildTable(RestRequest request, GetAliasesResponse response) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java]: Can you revert this change, I do not see any tests added that rely on the visibility being changed?

File: docs/reference/cat/alias.asciidoc [hunk 2]
```diff
@@ -54,5 +54,5 @@ alias4 test1 -      2            1,2
 The output shows that `alias2` has configured a filter, and specific routing
 configurations in `alias3` and `alias4`.
 
-If you only want to get information about a single alias, you can specify
-the alias in the URL, for example `/_cat/aliases/alias1`.
+If you only want to get information about a single or multiple aliases, you can specify
+the alias in the URL, for example `/_cat/aliases/alias1` or `/_cat/aliases/alias1,alias2`.
```
Comment (jasontedor) [inline docs/reference/cat/alias.asciidoc]: I wonder if rewording this to the following is simpler?


> If you only want to get information about specific aliases, you can specify the aliases in comma-delimited format as a URL parameter, e.g., `/_cat/aliases/aliases/alias1,alias2`.

File: rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml [hunk 3]
```diff
@@ -127,6 +127,37 @@
...
+  - do:
+        cat.aliases:
+            name: foo,bar
+
+  - match:
+      $body: /(^|\n)foo\s+test /
+
+  - match:
+      $body: /(^|\n)bar\s+test2 /
+
```
Comment (jasontedor) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml]: I think that this test should add a third index, and a third alias, and not request that alias (as you've already done) just so that we can be sure that a non-requested alias is not returned.

**Review by jasontedor — CHANGES_REQUESTED** [review]
Thanks @glefloch, this is a good start, I left some comments.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java (hunk 1 above)
Comment (glefloch) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java]: Yes I'm gonna revert this change

File: rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml (hunk 3 above)
Comment (glefloch) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml]: I will do that

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java [hunk 4]
```diff
@@ -46,7 +46,7 @@ public RestAliasAction(Settings settings, RestController controller) {
     @Override
     protected RestChannelConsumer doCatRequest(final RestRequest request, final NodeClient client) {
         final GetAliasesRequest getAliasesRequest = request.hasParam("alias") ?
-                new GetAliasesRequest(request.param("alias")) :
+                new GetAliasesRequest(request.param("alias").split(",")) :
```
Comment (olcbean) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java]: I think we should consider changing the parameter from `alias` to `aliases` as it is misleading to pack a `list` in `alias`.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java]: you could use `Strings.commaDelimitedListToStringArray` here as we do in other places.

**Comment by javanna** [comment]
jenkins please test this

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java (hunk 4 above)
Comment (glefloch) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestAliasAction.java]: I didn't knew about this class. I will use it.

**Comment by javanna** [comment]
retest this please

**Comment by glefloch** [comment]
@javanna the build has failed, but it seems like not tests are failed. Is there anything I can do to fix this?

**Comment by javanna** [comment]
@glefloch I think there may be a problem with backwards compatibility tests, as the new yaml test will not work against previous versions. It would need a skip section. I think we can take care of it when merging this in. @jasontedor are you ok with the state of this PR?

**Comment by jasontedor** [comment]
@javanna We saw similar build failures when we bumped versions in 5.x but not in master. I think that master needs to be merged into this PR branch to pick that up and then the PR CI can run again.

**Comment by javanna** [comment]
ok @glefloch can you merge master in please?

**Comment by glefloch** [comment]
@javanna I merged it

**Comment by javanna** [comment]
jenkins retest this please

**Comment by javanna** [comment]
@glefloch there's one test failure, as I mentioned above, around backwards compatibility. We run the yaml tests against previous versions to test backwards compatibility on the REST layer. Your updated test doesn't lead to the same result when run against previous versions that don't contain your fix. This is expected, could you add for now the following skip section to your newly added test please?

```
  - skip:
      version: " - 5.99.99"
      reason:  multiple aliases are supported only from 6.0.0 on

```

This way tests should go green and we can merge the PR. Later we will backport your fix and we will have to update the version in the skip section, both in master and 5.x. Let me know if you have any question about this.

**Comment by glefloch** [comment]
@javanna, I add the `skip` section. All tests passed locally.

**Comment by javanna** [comment]
jenkins retest this please

File: rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml [hunk 5]
```diff
@@ -127,6 +127,51 @@
...
+  - match:
+      $body: /(^|\n)[^bar\s+test3] /
```
Comment (jasontedor) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml]: I think we can and should make a stronger assertion here. We can request only the `alias` and `index` fields, request that the cat API sort the output on the `index` field, and then assert the *exact* expected output:

```
foo test
bar test2
```

**Review by jasontedor — CHANGES_REQUESTED** [review]
Almost there, thanks for all your work on this.

**Comment by glefloch** [comment]
@jasontedor  No problem ;) I'm going to do it.

File: rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml [hunk 6]
```diff
@@ -127,6 +127,57 @@
...
+  - match:
+      $body: |
+               /^  alias           \s+
+                   index
+                   \n
+                   foo             \s+
+                   test
+                   \n
+                   bar             \s+
+                   test2
```
Comment (jasontedor) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cat.aliases/10_basic.yaml]: I think this would be clearer if you put all the pieces that belong on a line together on the same line.

**Review by jasontedor — CHANGES_REQUESTED** [review]
One more thing. 😄

**Comment by jasontedor** [comment]
retest this please

**Review by jasontedor — APPROVED** [review]
LGTM.

**Review by javanna — APPROVED** [review]

**Comment by javanna** [comment]
Merged to master and cherry-picked to 5.x. Thanks a lot @glefloch

**Comment by jasontedor** [comment]
Indeed, thank you @glefloch and thank you @javanna for managing the integration.
